File: solutions/cpp/populating-next-right-pointers.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>
#include <queue>
#include <vector>

using std::cout;
using std::endl;
using std::queue;
using std::unique_ptr;
using std::vector;

template <typename T>
struct BinaryTreeNode {
  T data;
  unique_ptr<BinaryTreeNode<T>> left, right;
  BinaryTreeNode<T>* next;  // Populates this field.
};

void PopulateChildrenNextField(BinaryTreeNode<int>* start_node);

// @include
void ConstructRightSibling(BinaryTreeNode<int>* tree) {
  auto left_start = tree;
  while (left_start) {
    PopulateChildrenNextField(left_start);
    left_start = left_start->left.get();
  }
}

void PopulateChildrenNextField(BinaryTreeNode<int>* start_node) {
  auto iter = start_node;
  while (iter && iter->left) {
    iter->left->next = iter->right.get();
    if (iter->next) {
      iter->right->next = iter->next->left.get();
    }
    iter = iter->next;
  }
}
// @exclude

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1  7 4 6
  unique_ptr<BinaryTreeNode<int>> root = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{3, nullptr, nullptr, nullptr});
  root->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{2, nullptr, nullptr, nullptr});
  root->left->right = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{7, nullptr, nullptr, nullptr});
  root->left->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{1, nullptr, nullptr, nullptr});
  root->right = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{5, nullptr, nullptr, nullptr});
  root->right->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{4, nullptr, nullptr, nullptr});
  root->right->right = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{6, nullptr, nullptr, nullptr});
  ConstructRightSibling(root.get());
  assert(root->next == nullptr);
  assert(root->left->next == root->right.get());
  assert(root->left->left->next == root->left->right.get());
  assert(root->left->right->next == root->right->left.get());
  assert(root->right->left->next == root->right->right.get());
  return 0;
}

File: solutions/python/populating_next_right_pointers.py
class BinaryTreeNode:

    def __init__(self, data=None, left=None, right=None, nxt=None):
        self.data = data
        self.left = left
        self.right = right
        self.next = nxt  # Populates this field.


# @include
def construct_right_sibling(tree):
    def populate_children_next_field(start_node):
        while start_node and start_node.left:
            # Populate left child's next field.
            start_node.left.next = start_node.right
            # Populate right child's next field if iter is not the last node of
            # level.
            start_node.right.next = start_node.next and start_node.next.left
            start_node = start_node.next

    while tree and tree.left:
        populate_children_next_field(tree)
        tree = tree.left
# @exclude


def simple_test():
    #      3
    #    2   5
    root = BinaryTreeNode(3, BinaryTreeNode(2), BinaryTreeNode(5))
    construct_right_sibling(root)
    assert not root.next
    assert root.left.next is root.right
    assert not root.right.next


def main():
    simple_test()
    #      3
    #    2   5
    #  1  7 4 6
    root = BinaryTreeNode(3)
    root.left = BinaryTreeNode(2)
    root.left.right = BinaryTreeNode(7)
    root.left.left = BinaryTreeNode(1)
    root.right = BinaryTreeNode(5)
    root.right.left = BinaryTreeNode(4)
    root.right.right = BinaryTreeNode(6)
    construct_right_sibling(root)
    assert root.next is None
    assert root.left.next is root.right
    assert root.left.left.next is root.left.right
    assert root.left.right.next is root.right.left
    assert root.right.left.next is root.right.right


if __name__ == '__main__':
    main()
